src/main/tut/simple-algebraic-data-types.md

```diff
@@ -0,0 +1,366 @@
+```
+```tut:silent
+case class Element(
+  name: String,
+  symbol: String,
+  atomicNumber: Int
+)
+```
+...............
+```Haskell
+tupleToElem :: (String, String, Int) -> Element
+tupleToElem (n, s, a) = Element { name = n
+                                , symbol = s
+                                , atomicNumber = a }
+```
+```tut:silent
+val tupleToElem: (String, String, Int) => Element =
+  (n, s, a) => Element(n, s, a)
+```
+...............
+```Haskell
+elemToTuple :: Element -> (String, String, Int)
+elemToTuple e = (name e, symbol e, atomicNumber e)
+```
+```tut:silent
+val elemToTuple: Element => (String, String, Int) =
+  e => (e.name, e.symbol, e.atomicNumber)
+```
+...............
+```Haskell
+atomicNumber :: Element -> Int
+```
+```scala
+val atomicNumber: Element => Int
+```
+...............
+```Haskell
+startsWithSymbol :: Element -> Bool
+startsWithSymbol e = isPrefixOf (symbol e) (name e)
+```
+```tut:silent
+val startsWithSymbol: Element => Boolean =
+  e => e.name startsWith e.symbol
+```
+...............
+```Haskell
+startsWithSymbol e = symbol e `isPrefixOf` name e
+```
+```tut:silent
+val startsWithSymbol: Element => Boolean =
+  e => e.name startsWith e.symbol
+```
+...............
+```Haskell
+data Either a b = Left a | Right b
+```
+```tut:silent
+sealed trait Either[+A, +B]
+case class Left[+A](a: A) extends Either[A, Nothing]
+case class Right[+B](b: B) extends Either[Nothing, B]
+```
+...............
+```Haskell
+data OneOfThree a b c = Sinistral a | Medial b | Dextral c
+```
+```tut:silent
+sealed trait OneOfThree[A, B, C]
+case class Sinistral[A](a: A) extends OneOfThree[A, Nothing, Nothing]
+case class Medial[B](b: B) extends OneOfThree[Nothing, B, Nothing]
+case class Dextral[C](c: C) extends OneOfThree[Nothing, Nothing, C]
+```
+...............
+```Haskell
+Either a Void
+```
+```scala
+Either[A, Nothing]
+```
+...............
+```Haskell
+data Color = Red | Green | Blue
+```
+```tut:silent
+sealed trait Color
+case object Red extends Color
+case object Green extends Color
+case object Blue extends Color
+```
+...............
+```Haskell
+data Bool = True | False
+```
+```tut:silent
+sealed trait Bool
+case object True extends Bool
+case object False extends Bool
+```
+...............
+```Haskell
+data Maybe a = Nothing | Just a
+```
+```tut:silent
+sealed trait Maybe[+A]
+case object Nothin extends Maybe[Nothing]
+case class Just[+A](a: A) extends Maybe[A]
+```
+...............
+```Haskell
+data NothingType = Nothing
+```
+```tut:silent
+trait NothingType[Nothing]
+```
+...............
+```Haskell
+data JustType a = Just a
+```
+```scala
+sealed trait JustType[A]
+case class Just[A](a: A) extends JustType[A]
+```
+...............
+```Haskell
+data Maybe a = Either () a
+```
+```scala
+sealed trait Maybe[A]
+case class Either[A](u: Unit, a: A) extends Maybe[A]
+```
+...............
+```Haskell
+List a = Nil | Cons a (List a)
+```
+```tut:silent
+sealed trait List[+A]
```

Comment:
again, it the contravariance necessary?

Reply:
Not in this case, since we don't really do anything with `List` in this file/chapter. I just sorta tossed in the covariance annotation since it's convenient if somebody copy/pastes the snippet and tries to use it. Plus symmetry with `Either` and `Maybe` :p 

I'll remove it though, since it's really just adding noise here.